keps/sig-node/0028-node-readiness++.md

```diff
@@ -0,0 +1,237 @@
+
+
+#### Optional Extension - Track state of components along with Health
+
+Each of the individual node extension will most likely want to expose domain specific status for instrospection purposes.
+This status can be expressed as plugin specific CRDs which may be a good starting point.
+Down the line though, having a generic abstraction(s) to represent the Status of different plugins or extensions will be necessary to empower the wider introspection & remediation ecosystem.
+
+Assuming it is possible to define Status for some of the popular plugins like CSI, device plugins, kube proxy, logging & monitoring agents, the Lease API based approach can be extended as follows to easily track status of extensions in addition to Health.
+
+1. Embed the "Lease" API object in an extension specific Status object. An alternative is to have a local Lease object with the same name as the Status object.
+2. Similar to the workflow described above, the node controller, kubelet or extensions will register their status object as an extension via the Node.Spec.SystemExtensions field.
+3. The Node Controller will be updated to extract the Lease object (or identifying a separate Lease object by name in the local namespace) and influence health of extensions similar to the workflow described above.
+4. Ecosystem components like Monitoring, repair/remediation systems will be able to deterministically identify and expose or act on the status of the extensions.
+
+Here are some API illustations.
+
+##### Node Object extension
+
+```yaml
+type ExtensionFoo struct {
+ TypeMeta
+ ObjectMeta
+ Spec ExtensionFooStatus
+}
+
+type ExtensionFooStatus {
+  v1beta1.LeaseSpec
+  OtherFooStatus interface{} 
+}
+
+type Node struct {
+ …
+ Spec:
+  SystemExtensions []SystemExtension
+...
+}
+
+type SystemExtension struct {
+   Reference ObjectReference
+   AffectsReadiness bool // Ignore non-critical node extensions from node readiness
+}
+```
+
+##### Sample Extension Status Object
+
+```yaml
+type DevicePluginExtension struct {
+ TypeMeta
+ ObjectMeta
+ Status DevicePluginExtensionStatus
+}
+
+type DevicePluginExtensionStatus struct {
+  LeaseSpec
+  ComputeResource struct{ // embedded for brevity
+    Name: “foo.com/bar”
+    Quantity: x // Sample fields for illustration purposes
+    UnHealthy: y 
+    DeviceIDs []string
+  }
+}
+
+```
+
+One of the goals for this proposal is to evaluate the viability and usefullness of combining Status and Health for extensions.
+
+### Approach B - Use Readiness of Node extension pods
+
+This approach assumes that all node level extensions will get deployed as kubernetes pods and that the health of those extension pods can be exposed via existing Readiness probes.
+
+Similar to the previous approach `A`, a special taint will be used to track the overall readiness of the node and the node controller and kubelet will manage the special taint.
```

Comment:
Isn't this approach essentially a special case of Approach A, where the method of reporting readiness is done by looking at the specific pods? I guess in this case it doesn't have to write any per-component state - it is determined by the existing Pod state.

For those cases where pod state is enough, this seems really preferable to avoid additional, frequent writes to the API server. If we need to handle cases where pod state is insufficient, perhaps there is a way to allow the choice to use pod state or stored API objects on a per-extension basis.

Reply:
> Isn't this approach essentially a special case of Approach A

Yes. Are you proposing we do both `A` and `B`? 
As for handling the additional state, we can consider providing an extensible side car container that deals with health checks for extensions that are not easily modifiable. 
When it comes to tracking internal state of these extensions, a separate object will be necessary at some point though.

Reply:
Well, I am proposing that we make it easy for extension authors to indicate readiness using the existing control plane features whenever possible. Ideally that means that we don't need to write extension state, although if we do A then you can have a naive implementation that just writes it based on pod status.